pkg/hooks/manager.go

```diff
@@ -187,51 +187,59 @@ func (m *Manager) OnDefineDomain(domainSpec *virtwrapApi.DomainSpec, vmi *v1.Vir
 	if err != nil {
 		return "", fmt.Errorf("Failed to marshal domain spec: %v", domainSpec)
 	}
-	if callbacks, found := m.CallbacksPerHookPoint[hooksInfo.OnDefineDomainHookPointName]; found {
-		for _, callback := range callbacks {
-			if callback.Version == hooksV1alpha1.Version || callback.Version == hooksV1alpha2.Version {
-				vmiJSON, err := json.Marshal(vmi)
-				if err != nil {
-					return "", fmt.Errorf("Failed to marshal VMI spec: %v", vmi)
-				}
 
-				conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
-				if err != nil {
-					log.Log.Reason(err).Infof(dialSockErr, callback.SocketPath)
-					return "", err
-				}
-				defer conn.Close()
+	callbacks, found := m.CallbacksPerHookPoint[hooksInfo.OnDefineDomainHookPointName]
+	if !found {
+		return string(domainSpecXML), nil
+	}
 
-				ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
-				defer cancel()
+	if len(callbacks) == 0 {
+		return string(domainSpecXML), nil
+	}
 
-				switch callback.Version {
-				case hooksV1alpha1.Version:
-					client := hooksV1alpha1.NewCallbacksClient(conn)
-					result, err := client.OnDefineDomain(ctx, &hooksV1alpha1.OnDefineDomainParams{
-						DomainXML: domainSpecXML,
-						Vmi:       vmiJSON,
-					})
-					if err != nil {
-						return "", err
-					}
-					domainSpecXML = result.GetDomainXML()
-				case hooksV1alpha2.Version:
-					client := hooksV1alpha2.NewCallbacksClient(conn)
-					result, err := client.OnDefineDomain(ctx, &hooksV1alpha2.OnDefineDomainParams{
-						DomainXML: domainSpecXML,
-						Vmi:       vmiJSON,
-					})
-					if err != nil {
-						return "", err
-					}
-					domainSpecXML = result.GetDomainXML()
-				default:
-					panic("Should never happen, version compatibility check is done during Info call")
-				}
-			}
+	callback := callbacks[0]
+	if callback.Version == hooksV1alpha1.Version || callback.Version == hooksV1alpha2.Version {
+		panic("Should never happen, version compatibility check is done during Info call")
```

Comment:
In the original code this is the condition to run the loop iteration. I think you shouuld not panic here.

Reply:
I 100% agree.

By the way, @vasiliy-ul, WDYT of getting rid of this panic call altogether? It's very dangerous to have these in production code

Reply:
I think this panic was unreachable in the original code since the `switch` was inside the condition `if callback.Version == hooksV1alpha1.Version || callback.Version == hooksV1alpha2.Version {`. So yeah, removing it should do no harm.

Reply:
Right.

I don't like this habit :smile: basically we use it to signal ourselves that a code that we thought was unreachable is in fact reachable. That's great to be aware of that, but panic is an overkill and could harm production code. I would personally replace it with some log entry.

@VirrageS WDYT? Do you agree? Is it possible for you to change that?

Reply:
Sure, sounds good to me. I just didn't want to change any logic but if you feel like this is the right call, I'm defenitely going to change it :))

Reply:
I've changed it to `log.Log.Errorf`.